### PR TITLE
feat(argo-cd): enabled server cluster role permissions overriding

### LINKED ISSUE #2605: Enhancement: Support for Overriding Argo CD Server ClusterRole Permissions
### Is your feature request related to a problem?

When deploying Argo CD using the helm chart, enabling cluster roles for the Argo CD server results in the server obtaining extensive permissions due to the broadly defined ClusterRole. This broad permission set can be a concern in environments where security policies require minimum necessary permissions. Moreover, if there's a need to reduce the cluster role permissions for the application controller, it becomes challenging to reconcile and adjust the permissions for the server's ClusterRole accordingly. 

The concern is primarily around the default ClusterRole permissions which are as follows:

```yaml
{{- if .Values.createClusterRoles }}
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: {{ include "argo-cd.server.fullname" . }}
  labels:
    {{- include "argo-cd.labels" (dict "context" . "component" .Values.server.name "name" .Values.server.name) | nindent 4 }}
rules:
  - apiGroups:
      - '*'
    resources:
      - '*'
    verbs:
      - delete
      - get
      - patch
  - apiGroups:
      - ""
    resources:
      - events
    verbs:
      - list
      {{- if (index .Values.configs.params "application.namespaces") }}
      - create
      {{- end }}
  - apiGroups:
      - ""
    resources:
      - pods
      - pods/log
    verbs:
      - get
  {{- if eq (toString (index .Values.configs.cm "exec.enabled")) "true" }}
  - apiGroups:
      - ""
    resources:
      - pods/exec
    verbs:
      - create
  {{- end }}
  - apiGroups:
      - argoproj.io
    resources:
      - applications
      - applicationsets
    verbs:
      - get
      - list
      - update
      - watch
  - apiGroups:
      - batch
    resources:
      - jobs
    verbs:
      - create
  - apiGroups:
      - argoproj.io
    resources:
      - workflows
    verbs:
      - create
{{- end }}
```

### Related helm chart

argo-cd

### Describe the solution you'd like

An option to override or customize the ClusterRole permissions for the Argo CD server would provide flexibility to adjust permissions in line with organizational security policies. A `values.yaml` entry to specify custom resource permissions or completely override the default cluster role would be desirable.

For example, introducing `.Values.server.clusterRole.rules` could allow users to specify their own sets of permissions directly in the `values.yaml` file.
```yaml
{{- if .Values.createClusterRoles }}
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: {{ include "argo-cd.server.fullname" . }}
  labels:
    {{- include "argo-cd.labels" (dict "context" . "component" .Values.server.name "name" .Values.server.name) | nindent 4 }}
rules:
  {{- if .Values.server.clusterRoleRules.enabled }}
    {{- toYaml .Values.server.clusterRoleRules.rules | nindent 2 }}
  {{- else }}
  - apiGroups:
      - '*'
    resources:
      - '*'
    verbs:
      - delete
      - get
      - patch
  - apiGroups:
      - ""
    resources:
      - events
    verbs:
      - list
      {{- if (index .Values.configs.params "application.namespaces") }}
      - create
      {{- end }}
  - apiGroups:
      - ""
    resources:
      - pods
      - pods/log
    verbs:
      - get
  {{- if eq (toString (index .Values.configs.cm "exec.enabled")) "true" }}
  - apiGroups:
      - ""
    resources:
      - pods/exec
    verbs:
      - create
  {{- end }}
  - apiGroups:
      - argoproj.io
    resources:
      - applications
      - applicationsets
    verbs:
      - get
      - list
      - update
      - watch
  - apiGroups:
      - batch
    resources:
      - jobs
    verbs:
      - create
  - apiGroups:
      - argoproj.io
    resources:
      - workflows
    verbs:
      - create
  {{- end }}
{{- end }}
```

### Describe alternatives you've considered

As a workaround, manually editing the ClusterRole after deployment is possible but not ideal as it interferes with GitOps principles and requires additional operational steps that could potentially be automatable.

### Additional context

This feature request aims to increase the security adaptability of the Argo CD helm chart for use in various environments with different security postures. Being able to fine-tune permissions would enhance its usability and safety.

### PR DESCRIPTION
**Pull request by edmondshtogu**
Resolves #2605 

---

Checklist:

* [x] I have bumped the chart version according to [versioning](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#versioning)
* [x] I have updated the documentation according to [documentation](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#documentation)
* [x] I have updated the chart changelog with all the changes that come with this pull request according to [changelog](https://github.com/argoproj/argo-helm/blob/main/CONTRIBUTING.md#changelog).
* [x] Any new values are backwards compatible and/or have sensible default.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md).
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/stable/developer-guide/ci/)).

### CODE REVIEW

File: charts/argo-cd/templates/argocd-server/clusterrole.yaml [hunk 1]
```diff
@@ -6,6 +6,9 @@ metadata:
   labels:
     {{- include "argo-cd.labels" (dict "context" . "component" .Values.server.name "name" .Values.server.name) | nindent 4 }}
 rules:
+  {{- if .Values.server.clusterRoleRules.enabled }}
+    {{- toYaml .Values.server.clusterRoleRules.rules | nindent 2 }}
+  {{- else }}
```
Comment (yu-croco) [inline charts/argo-cd/templates/argocd-server/clusterrole.yaml:11]: We can use `rules` like below, so maybe we don't need `enabled` flag ? 🙋 

```yaml
  {{- if .Values.server.clusterRoleRules.rules }}
    {{- toYaml .Values.server.clusterRoleRules.rules | nindent 2 }}
  {{- else }}
...
```

Comment (yu-croco) [reply charts/argo-cd/templates/argocd-server/clusterrole.yaml:11]: Oh... 🙄

**Comment by edmondshtogu** [comment]
Hi guys, can someone review and unblock this PR so we can merge the enhancement if there is no further change required?

**Review by yu-croco — APPROVED** [review]
Thank you for your contribution. LGTM.

**Review by mkilchhofer — APPROVED** [review]

**Pull request merged**